### PR TITLE
[1429] snc serializer part 4

### PR DESCRIPTION
**Pull request by defong**
### Context
search and compare serialiizer part 4
site/campus and related touch points

### Changes proposed in this pull request
Added course sites serializer
Added application open date
Added has vacanies

### Guidance to review
Its going to get mess from henceforth, PR to make it better is welcome

part 4 ot x

Refer to the almost complete
https://github.com/DFE-Digital/manage-courses-backend/pull/377

For the full whack

FYI 

Part 1
https://github.com/DFE-Digital/manage-courses-backend/pull/381

Part 2
https://github.com/DFE-Digital/manage-courses-backend/pull/387

Part 3
https://github.com/DFE-Digital/manage-courses-backend/pull/390

Probably easier to follow the commits

### Checklist

- [ ] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[[L] Serialize course to sync with search-and-compare](https://trello.com/c/EVoUeZHg/1429-l-serialize-course-to-sync-with-search-and-compare)

File: app/serializers/search_and_compare/course_serializer.rb [hunk 1]
```diff
@@ -63,5 +62,53 @@ def get_subjects
...
+        {
+          **campus_default_value,
+          VacStatus: site_status.vac_status_before_type_cast,
+          Name: site_status.site.location_name,
+          CampusCode: site_status.site.code,
+          Location: { **location_default_value, Address: address }
+        }
+      end
+    end
+    #end end end
```
Comment (misaka) [inline app/serializers/search_and_compare/course_serializer.rb]: END IT!!

Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: i know, just approve this so i can get on to making 3 more pr to end this nightmare all together

File: app/serializers/search_and_compare/course_serializer.rb [hunk 2]
```diff
@@ -63,5 +62,53 @@ def get_subjects
         }
       end
     end
+
+    # Campuses_related_Mapping
+    attribute(:Campuses)                              { get_campuses }
+    # using server time not utc, so it's local time?
+    attribute(:ApplicationsAcceptedFrom)              { object.applications_open_from.to_date.strftime('%Y-%m-%dT%H:%M:%S') }
+    attribute(:HasVacancies)                          { object.has_vacancies? }
+    def get_campus_default_value
```
Comment (misaka) [inline app/serializers/search_and_compare/course_serializer.rb]: In Ruby we don't typically use `get_` to signal a getter ... this ain't C#!

Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: Also, I think these should be private.

File: app/serializers/search_and_compare/course_serializer.rb [hunk 3]
```diff
@@ -63,5 +62,53 @@ def get_subjects
...
+        Id: 0,
+        FormattedAddress: nil,
+        GeoAddress: nil,
+        Latitude: nil,
+        Longitude: nil,
+        LastGeocodedUtc: '0001-01-01T00:00:00'
+      }
+    end
+
+    def get_address(address1:, address2:, address3:, address4:, postcode:)
```
Comment (dankmitchell) [inline app/serializers/search_and_compare/course_serializer.rb]:
```
def full_address
    [address1, address2, address3, address4, postcode].select(&:present?).join("<br> ").html_safe
end
```

?

File: app/serializers/search_and_compare/course_serializer.rb (hunk 2 above)
Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: `campus_default_value` ?

File: app/serializers/search_and_compare/course_serializer.rb (hunk 3 above)
Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: cannot use html related bits as it needs to be feed in to geocoder

`reject` is always better then `select`

`present?` is that true for `""`?

File: app/serializers/search_and_compare/course_serializer.rb (hunk 2 above)
Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: Ya, that's the way we usually do it, ya ... call it what it is. :)

File: app/serializers/search_and_compare/course_serializer.rb (hunk 3 above)
Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: renamed method to `full_address`

its going to be needed for provider location and contact details address

Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: Why is `reject` better than `select`?

Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: just a passing motion, 

select is opposite to reject
present is opposite to blank

ruby is about the only one that has the opposite named method to everything


> array reject blank join

> array select present join

either one is fine by me, favouring `reject` i can only why....

File: spec/serializers/search_and_compare/course_serializer_spec.rb [hunk 4]
```diff
@@ -18,6 +18,41 @@
...
+              address2: '',
+              address3: '',
+              address4: '',
+              postcode: ''
+            },
+            site_save_validate: false
+          },
+        ]
+        { site_statuses_and_sites: site_statuses_and_sites }
+      end
```
Comment (misaka) [inline spec/serializers/search_and_compare/course_serializer_spec.rb]: Is this kind of meta-representation of our test data really worth it? tbh, I've been wanting to find time to remove this kind of thing and replace it with just plain factory generation, so unless it's really ugly to do if with the existing factory system, I'd rather we just did that.

File: app/serializers/search_and_compare/course_serializer.rb (hunk 3 above)
Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: wat?

File: app/serializers/search_and_compare/course_serializer.rb [hunk 5]
```diff
@@ -23,45 +23,98 @@ class CourseSerializer < ActiveModel::Serializer
...
     # Subjects_related_Mapping
     attribute(:IsSen)                                 { object.is_send? }
-    attribute(:CourseSubjects)                        { get_subjects }
+    attribute(:CourseSubjects)                        { course_subjects }
+
+
+    # Campuses_related_Mapping
+    attribute(:Campuses)                              { campuses }
+    # using server time not utc, so it's local time?
+    attribute(:ApplicationsAcceptedFrom)              { object.applications_open_from.to_date.strftime('%Y-%m-%dT%H:%M:%S') }
```
Comment (misaka) [inline app/serializers/search_and_compare/course_serializer.rb]: Not clear on what's going on here ... `applications_open_from` is a date, does search-and-compare store it as a time?

File: app/serializers/search_and_compare/course_serializer.rb (hunk 1 above)
Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: Do we need that line though? Or is this a counter to remind you how many more PRs you need to make?

File: spec/serializers/search_and_compare/course_serializer_spec.rb (hunk 4 above)
Comment (defong) [reply spec/serializers/search_and_compare/course_serializer_spec.rb]: I think you mentioning a number of things here.
`meta-representation`, I think its a good thing, ie grab another json, elicit the bits, put it into a course model and related models.

`plain factory generation/factory system`, I don't feel its applicable for this task as the default approach, as the task deals with actual stored data

data stored previously valid but currently it is invalid

eventually `plain factory generation/factory system` should be the the way to go

Comment (defong) [reply spec/serializers/search_and_compare/course_serializer_spec.rb]: https://github.com/DFE-Digital/manage-courses-backend/pull/403

I think this shows the folly of `plain factory generation` I could of omitted 2 of the 3 values that drives the 6 values. and the tests would of still passed

File: app/serializers/search_and_compare/course_serializer.rb (hunk 3 above)
Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: > > array reject blank join

```array.reject(&:blank?).join('/n')```
 
> > array select present join

```array.select(&:present?).join("<br> ") ```

File: app/serializers/search_and_compare/course_serializer.rb (hunk 5 above)
Comment (defong) [reply app/serializers/search_and_compare/course_serializer.rb]: input
`'2018-10-09T00:00:00'`

expected output 
`'2018-10-09T00:00:00'`

actual output
`'2018-10-09T00:00:00Z'`

Notice the `Z` for the time zone

Comment (misaka) [reply app/serializers/search_and_compare/course_serializer.rb]: I just checked and it is a `timestamp` (no tz :( ).

File: spec/serializers/search_and_compare/course_serializer_spec.rb (hunk 4 above)
Comment (misaka) [reply spec/serializers/search_and_compare/course_serializer_spec.rb]: I'm going to take a crack at this tonight. I'm still a bit reluctant to put in this kind of test generation stuff.

Comment (defong) [reply spec/serializers/search_and_compare/course_serializer_spec.rb]: thanks @misaka will continue to send my rubbish code to you

**Pull request merged**